clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -0,0 +1,440 @@
+          else {
+            NewBounds = std::max(NewBounds, EB->Gen[succ][V]);
+            EB->Gen[succ][V] = NewBounds;
+          }
+        }
+
+    // For conditions of the form "if (*(p + i))"
+    } else if (const auto *BO = dyn_cast<BinaryOperator>(Exp)) {
+      if (BO->getOpcode() != BO_Add)
+        return;
+
+      const auto *D =
+        dyn_cast<DeclRefExpr>(IgnoreCasts(BO->getLHS()));
+      const auto *Lit = dyn_cast<IntegerLiteral>(IgnoreCasts(BO->getRHS()));
+      if (!D || !Lit)
+        return;
+
+      if (const auto *V = dyn_cast<VarDecl>(D->getDecl())) {
+        if (V->getType()->isCheckedPointerNtArrayType()) {
+          unsigned NewBounds = Lit->getValue().getLimitedValue();
+          if (!EB->Gen[succ].count(V))
+            EB->Gen[succ].insert(std::make_pair(V, NewBounds));
+          else {
+            NewBounds = std::max(NewBounds, EB->Gen[succ][V]);
+            EB->Gen[succ][V] = NewBounds;
+          }
+        }
+      }
+    }
+  }
+}
+
+void BoundsAnalysis::ComputeKillSets(BlockMapTy BlockMap) {
+  // For a block B, a variable v is added to Kill[B] if v is assigned to in B.
+
+  for (const auto B : BlockMap) {
+    auto EB = B.second;
+    DeclSetTy DefinedVars;
+
+    for (auto Elem : *(EB->Block))
+      if (Elem.getKind() == CFGElement::Statement)
+        CollectDefinedVars(Elem.castAs<CFGStmt>().getStmt(), DefinedVars);
+
+    for (const auto V : DefinedVars)
+      EB->Kill.insert(V);
+  }
+}
+
+void BoundsAnalysis::CollectDefinedVars(const Stmt *S, DeclSetTy &DefinedVars) {
+  if (!S)
+    return;
+
+  Expr *E = nullptr;
+  if (const auto *UO = dyn_cast<const UnaryOperator>(S)) {
+    if (UO->isIncrementDecrementOp())
+      E = IgnoreCasts(UO->getSubExpr());
+  } else if (const auto *BO = dyn_cast<const BinaryOperator>(S)) {
+    if (BO->isAssignmentOp())
+      E = IgnoreCasts(BO->getLHS());
+  }
+
+  if (E) {
+    if (const auto *D = dyn_cast<DeclRefExpr>(E))
+      if (const auto *V = dyn_cast<VarDecl>(D->getDecl()))
+        if (V->getType()->isCheckedPointerNtArrayType())
+          DefinedVars.insert(V);
+  }
+
+  for (const auto I : S->children())
+    CollectDefinedVars(I, DefinedVars);
+}
+
+void BoundsAnalysis::ComputeInSets(ElevatedCFGBlock *EB, BlockMapTy BlockMap) {
+  // In[B1] = n Out[B*->B1], where B* are all preds of B1.
+
+  BoundsMapTy Intersections;
+  bool ItersectionEmpty = true;
+
+  for (const CFGBlock *pred : EB->Block->preds()) {
+    if (SkipBlock(pred))
+      continue;
+
+    auto PredEB = BlockMap[pred];
+
+    if (ItersectionEmpty) {
+      Intersections = PredEB->Out[EB->Block];
+      ItersectionEmpty = false;
+    } else
+      Intersections = Intersect(Intersections, PredEB->Out[EB->Block]);
+  }
+
+  EB->In = Intersections;
+}
+
+void BoundsAnalysis::ComputeOutSets(ElevatedCFGBlock *EB,
+                                    BlockMapTy BlockMap,
+                                    WorkListTy &WorkList) {
+  // Out[B1->B2] = In[B1] u Gen[B1->B2].
```

Comment:
I think you should do the widening of bounds here, not with a separate function as you are doing.  The widening should happen along the true branch of an if statement that checks the value stored at the (potentially widened) upper bound of an nt_array_ptr variable.   I recommend something like the following computation:
1. Let Tmp  = In[B1].   Remove (kill) any entries in Tmp for any variable v with bounds(e1, e2),  where v or a variable in e1 or e2 is assigned to in B1.
2. For successors s of B1, let Out[B1->S] = Tmp.
3. Widen: If  B1 ends in the form "if (*(v + k))"  (where k could = 0, i.e. if (*v)). and v is an nt_array_ptr,
- Compute the upper bound u of v.  If v has no entry in  Tmp, and the declared upper bound of v is u', let u = u'.   If v has an entry in Tmp, and the declared bounds are u', where u', has the form (p + j), let u = p + (j + Tmp[v]).
- If u has the form v + k, widen the bounds to v + (k+1), i.e. set Tmp[v] = k+1.
- Let B2 be the true branch of "if (*(v + k))".  Let Out[B1->B2] = Tmp.

Reply:
The analysis is edge-based. I need to handle sequential access of the `_Nt_array_ptr`'s.
For example,

`if (*(p)) //bounds: 1`
` if (*(p+1)) //bounds: 2`
 `if (*(p+2)) //bounds: 3`

`if (*(p)) //bounds: 1`
 `if (*(p+2)) //bounds: 1`
` if (*(p+3)) //bounds: 1`